### PR TITLE
BUG/TST: PeriodArray.__setitem__ with slice and list-like value

### LINKED ISSUE #23978: BUG/TST: PeriodArray.__setitem__ with slice and list-like value fails
```python

In [1]: import pandas as pd; import numpy as np

In [2]: arr = pd.core.arrays.period_array([2000, 2001], 'D')

In [3]: arr[:1] = ['2001']
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-907919e36fd9> in <module>()
----> 1 arr[:1] = ['2001']

~/sandbox/pandas/pandas/core/arrays/period.py in __setitem__(self, key, value)
    358         # work, since the freq can't be inferred.
    359         if is_list_like(value):
--> 360             if len(key) != len(value) and not com.is_bool_indexer(key):
    361                 msg = ("shape mismatch: value array of length '{}' does not "
    362                        "match indexing result of length '{}'.")

TypeError: object of type 'slice' has no len()

```

---

We should be checking that the if `key` is a slice

```python
            is_slice = isinstance(key, slice)
            if (not is_slice
                    and len(key) != len(value)
                    and not com.is_bool_indexer(key)):
                msg = ("shape mismatch: value array of length '{}' does not "
                       "match indexing result of length '{}'.")
                raise ValueError(msg.format(len(key), len(value)))
            if not is_slice and len(key) == 0:
                return
```

Then, if they slice length doesn't match with value, NumPy will complain at the end when we do `self._data[key] = value`.

I've fixed this on my all-in-one branch. To close this issue, we should also add a base extension tests to BaseSetitemTests like

```python
def test_setitem_slice_mismatch_length_raises(self, data):
    arr = data[:5]
    with pytest.raises(ValueError):
        arr[:1] = arr[:2]
```

and a successful one like

```python
def test_setitem_slice_array(self, data):
    arr = data[:5].copy()
    arr[:5] = data[-5:]
    self.assert_extension_array_equal(arr, data[-5:])
```

(those are untested)

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [ ] closes #23978 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by TomAugspurger — COMMENTED** [review]
A few things

1. The tests should be written as base extension tests, see http://pandas-docs.github.io/pandas-docs-travis/extending.html#testing-extension-arrays. These can go in tests/extension/base/setitem.py::BaseSetitemTests.
2. Extension Arrays define a `data` fixture. You don't need to define one.
3. Once you've written the test, `pytest pandas/tests/extension/test_period.py::TestSetitem` should fail, because of the bug from the original issue.

**Comment by TomAugspurger** [comment]
Great, your test has successfully raised the original bug: https://travis-ci.org/pandas-dev/pandas/jobs/461252693#L2863

So now you'll need to make the changes to `PeriodArray.__setitem__` to fix the bug. Make sense?

**Comment by charlesdong1991** [comment]
thank you very much @TomAugspurger ! this time it should work!

**Comment by TomAugspurger** [comment]
Some linting failures

https://travis-ci.org/pandas-dev/pandas/jobs/461404542#L2851

https://travis-ci.org/pandas-dev/pandas/jobs/461404542#L2868

**Comment by charlesdong1991** [comment]
HI, @TomAugspurger , ehh, this Flaky error appears again... Is it because my changes didn't pass the hinting test?

File: pandas/tests/extension/test_period.py [hunk 1]
```diff
@@ -147,7 +148,16 @@ class TestReshaping(BasePeriodTests, base.BaseReshapingTests):
 
 
 class TestSetitem(BasePeriodTests, base.BaseSetitemTests):
-    pass
+
+    def test_setitem_slice_mismatch_length_raises(self, data):
```
Comment (TomAugspurger) [inline pandas/tests/extension/test_period.py]: Ahh, just noticed. This should be on the base class BaseSetitemTests. Then all the subclasses will be tested, including period.

Comment (TomAugspurger) [reply pandas/tests/extension/test_period.py]: Same for the other test.

Comment (charlesdong1991) [reply pandas/tests/extension/test_period.py]: ahh, i see, yeah, makes sense!! thx!

File: pandas/tests/extension/test_period.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import pandas as pd
 from pandas.core.arrays import PeriodArray
 from pandas.tests.extension import base
+import pandas.util.testing as tm
```
Comment (TomAugspurger) [inline pandas/tests/extension/test_period.py]: Probably delete this

File: pandas/tests/extension/test_period.py [hunk 3]
```diff
@@ -147,6 +148,7 @@ class TestReshaping(BasePeriodTests, base.BaseReshapingTests):
 
 
 class TestSetitem(BasePeriodTests, base.BaseSetitemTests):
+
```
Comment (TomAugspurger) [inline pandas/tests/extension/test_period.py]: And this.

**Review by jreback — CHANGES_REQUESTED** [review]
needs a whatsnew entry

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @charlesdong1991

**Pull request merged**